google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionGroup.java

```diff
@@ -101,4 +109,60 @@ public void getPartitions(
 
     observer.onCompleted();
   }
+
+  public ApiFuture<List<QueryPartition>> getPartitions(long desiredPartitionCount) {
+    Preconditions.checkArgument(
+        desiredPartitionCount > 0, "Desired partition count must be one or greater");
+
+    // Partition queries require explicit ordering by __name__.
+    final Query queryWithDefaultOrder = orderBy(FieldPath.DOCUMENT_ID);
+
+    if (desiredPartitionCount == 1) {
+      // Short circuit if the user only requested a single partition.
+      return ApiFutures.immediateFuture(
+          Collections.singletonList(new QueryPartition(queryWithDefaultOrder, null, null)));
+    } else {
+      PartitionQueryRequest.Builder request = PartitionQueryRequest.newBuilder();
+      request.setStructuredQuery(queryWithDefaultOrder.buildQuery());
+      request.setParent(options.getParentPath().toString());
+
+      // Since we are always returning an extra partition (with en empty endBefore cursor), we
+      // reduce the desired partition count by one.
+      request.setPartitionCount(desiredPartitionCount - 1);
+
+      final TraceUtil traceUtil = TraceUtil.getInstance();
+      Span span = traceUtil.startSpan(TraceUtil.SPAN_NAME_PARTITIONQUERY);
+      try (Scope scope = traceUtil.getTracer().withSpan(span)) {
+        return ApiFutures.transform(
+            rpcContext.sendRequest(
+                request.build(), rpcContext.getClient().partitionQueryPagedCallable()),
+            new ApiFunction<PartitionQueryPagedResponse, List<QueryPartition>>() {
+              @Override
+              public List<QueryPartition> apply(PartitionQueryPagedResponse input) {
+                ImmutableList.Builder<QueryPartition> partitions = ImmutableList.builder();
+                Object[] lastCursor = null;
+                for (Cursor cursor : input.iterateAll()) {
+                  Object[] decodedCursorValue = new Object[cursor.getValuesCount()];
+                  for (int i = 0; i < cursor.getValuesCount(); ++i) {
+                    decodedCursorValue[i] =
+                        UserDataConverter.decodeValue(rpcContext, cursor.getValues(i));
+                  }
+                  partitions.add(
+                      new QueryPartition(queryWithDefaultOrder, lastCursor, decodedCursorValue));
+                  lastCursor = decodedCursorValue;
+                }
+                partitions.add(new QueryPartition(queryWithDefaultOrder, lastCursor, null));
+
+                return partitions.build();
+              }
+            },
+            MoreExecutors.directExecutor());
+      } catch (ApiException exception) {
+        span.setStatus(Status.UNKNOWN.withDescription(exception.getMessage()));
+        throw FirestoreException.apiException(exception);
+      } finally {
+        span.end(TraceUtil.END_SPAN_OPTIONS);
+      }
```

Comment:
FWIW these Codecov warnings seem incorrect and make it hard to read this code...

Reply:
Agreed, not sure why they're showing up for this change. There is a test added below that is running this code.